File: architecture-skills/design-patterns/examples/composite/composite.go
// Composite pattern in Go.
//
// Files and directories satisfy one small Node interface; a directory
// holds a []Node slice, so callers recurse without type switches.

package main

import (
	"fmt"
	"strings"
)

type Node interface {
	Size() int64
	Print(indent int)
}

type File struct {
	Name  string
	Bytes int64
}

func (f File) Size() int64 { return f.Bytes }

func (f File) Print(indent int) {
	fmt.Printf("%s%s (%d B)\n", strings.Repeat(" ", indent), f.Name, f.Bytes)
}

type Directory struct {
	Name     string
	Children []Node
}

func (d *Directory) Add(child Node) *Directory {
	d.Children = append(d.Children, child)
	return d
}

func (d *Directory) Size() int64 {
	var total int64
	for _, child := range d.Children {
		total += child.Size()
	}
	return total
}

func (d *Directory) Print(indent int) {
	fmt.Printf("%s%s/ (%d B)\n", strings.Repeat(" ", indent), d.Name, d.Size())
	for _, child := range d.Children {
		child.Print(indent + 2)
	}
}

func main() {
	src := (&Directory{Name: "src"}).
		Add(File{Name: "main.py", Bytes: 1200}).
		Add(File{Name: "util.py", Bytes: 450})
	docs := (&Directory{Name: "docs"}).Add(File{Name: "readme.md", Bytes: 800})

	project := (&Directory{Name: "project"}).
		Add(src).
		Add(docs).
		Add(File{Name: ".gitignore", Bytes: 20})

	project.Print(0)
}

File: architecture-skills/design-patterns/examples/composite/composite.py
"""Composite pattern in Python.

Files and directories share one interface, so callers can total sizes
or print trees without caring whether a node is a leaf or a container.
Duck typing would work without the ABC; it is kept as documentation.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class Node(ABC):
    def __init__(self, name: str) -> None:
        self.name = name

    @abstractmethod
    def size(self) -> int:
        """Total size in bytes."""

    @abstractmethod
    def print_tree(self, indent: int = 0) -> None: ...


class File(Node):
    def __init__(self, name: str, size: int) -> None:
        super().__init__(name)
        self._size = size

    def size(self) -> int:
        return self._size

    def print_tree(self, indent: int = 0) -> None:
        print(f"{' ' * indent}{self.name} ({self._size} B)")


class Directory(Node):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self._children: list[Node] = []

    def add(self, child: Node) -> Directory:
        self._children.append(child)
        return self

    def size(self) -> int:
        return sum(child.size() for child in self._children)

    def print_tree(self, indent: int = 0) -> None:
        print(f"{' ' * indent}{self.name}/ ({self.size()} B)")
        for child in self._children:
            child.print_tree(indent + 2)


if __name__ == "__main__":
    project = (
        Directory("project")
        .add(Directory("src").add(File("main.py", 1200)).add(File("util.py", 450)))
        .add(Directory("docs").add(File("readme.md", 800)))
        .add(File(".gitignore", 20))
    )
    project.print_tree()
